base_layer/core/src/mempool/utx_pool/utx_pool_storage.rs

```diff
@@ -0,0 +1,165 @@
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    blocks::block::Block,
+    mempool::utx_pool::{PrioritizedTransaction, UTxPoolConfig, UTxPoolError, UTxPriority},
+    transaction::Transaction,
+    types::Signature,
+};
+use std::{
+    collections::{BTreeMap, HashMap},
+    convert::TryFrom,
+};
+
+/// UTxPool makes use of UTxPoolStorage to provide thread save access to its Hashmap and BTreeMap.
+pub struct UTxPoolStorage {
+    config: UTxPoolConfig,
+    utx_hmap: HashMap<Signature, PrioritizedTransaction>,
+    utx_tree: BTreeMap<UTxPriority, Signature>,
+}
+
+impl UTxPoolStorage {
+    /// Create a new UTxPoolStorage with the specified configuration
+    pub fn new(config: UTxPoolConfig) -> Self {
+        Self {
+            config,
+            utx_hmap: HashMap::new(),
+            utx_tree: BTreeMap::new(),
+        }
+    }
+
+    fn lowest_priority(&self) -> &UTxPriority {
+        self.utx_tree.iter().next().unwrap().0
+    }
+
+    fn remove_lowest_priority_utx(&mut self) {
+        if let Some((priority, sig)) = self.utx_tree.iter().next().map(|(p, s)| (p.clone(), s.clone())) {
+            self.utx_hmap.remove(&sig);
+            self.utx_tree.remove(&priority);
+        }
+    }
+
+    /// Insert a new transaction into the UTxPoolStorage. Low priority transactions will be removed to make space for
+    /// higher priority transactions. The lowest priority transactions will be removed when the maximum capacity is
+    /// reached and the new transaction has a higher priority than the currently stored lowest priority transaction.
+    pub fn insert_utx(&mut self, utx: Transaction) -> Result<(), UTxPoolError> {
+        if utx.body.kernels.len() == 0 {
+            return Err(UTxPoolError::NoKernelsFound);
+        }
+        let utx_key = utx.body.kernels[0].excess_sig.clone();
+        if !self.utx_hmap.contains_key(&utx_key) {
```

Comment:
If the transaction is already in the pool, you return `Ok(())`. Shouldn't this be an Error?

Reply:
I was trying to make the insert function idempotent. I thought it should only fail if there is a storage or thread issue. I can change this if you think it should rather return an Error

Reply:
Idempotent is just fine -- assuming the tx_key is the hash of the transaction, right?. Could you mention this behaviour in the docstring?